File: osm_extract_simplify.py
# coding: utf-8

import argparse
import copy
import gzip
import json
import os
import sys
from collections import defaultdict
from dataclasses import dataclass, field

import osmium

if sys.stdout.encoding and sys.stdout.encoding.lower().replace("-", "") != "utf8":
    sys.stdout.reconfigure(encoding="utf-8", errors="replace")


@dataclass
class RoadSegment:
    fclass: str       # "motorway", "trunk", "primary", "secondary"
    name: str         # 道路名（""=未設定）
    ref: str          # 路線番号（""=未設定）
    points: list = field(default_factory=list)  # [(lon, lat), ...]


TARGET_CLASSES = ["motorway", "trunk", "primary", "secondary"]


class RoadHandler(osmium.SimpleHandler):
    def __init__(self, target_classes):
        super().__init__()
        self.segments = []
        self._targets = set(target_classes)
        self._link_targets = {c + "_link" for c in target_classes}

    def way(self, w):
        highway = w.tags.get("highway", "")
        if highway in self._targets:
            fclass = highway
        elif highway in self._link_targets:
            fclass = highway[:-5]  # strip "_link"
        else:
            return

        try:
            points = [(n.lon, n.lat) for n in w.nodes]
        except osmium.InvalidLocationError:
            return

        if len(points) < 2:
            return

        self.segments.append(RoadSegment(
            fclass=fclass,
            name=w.tags.get("name", ""),
            ref=w.tags.get("ref", ""),
            points=points,
        ))


def read_pbf(path, target_classes):
    handler = RoadHandler(target_classes)
    handler.apply_file(path, locations=True, idx="flex_mem")
    return handler.segments


def interpolate_attributes(seg_i, seg_j):
    """隣接するセグメント間で name/ref を補完する。"""
    for src, dst in [(seg_i, seg_j), (seg_j, seg_i)]:
        if src.name != "" and src.name == dst.name:
            if dst.ref == "" and src.ref != "":
                dst.ref = src.ref
        if src.ref != "" and src.ref == dst.ref:
            if dst.name == "" and src.name != "":
                dst.name = src.name


def are_adjacent(seg_a, seg_b):
    """2つのセグメントの端点が接続しているか判定する。"""
    return (seg_a.points[-1] == seg_b.points[0] or
            seg_a.points[0] == seg_b.points[-1])


def combine_segments(segments, max_iter=3):
    for iteration in range(max_iter):
        print(f"iteration {iteration}")

        # 端点からセグメントインデックスへの辞書を構築（O(n)）
        start_map = defaultdict(set)  # (fclass, 始点座標) -> {index, ...}
        end_map = defaultdict(set)    # (fclass, 終点座標) -> {index, ...}
        for idx, seg in enumerate(segments):
            start_map[(seg.fclass, seg.points[0])].add(idx)
            end_map[(seg.fclass, seg.points[-1])].add(idx)

        new_segments = []
        removed = set()

        for i in range(len(segments)):
            if i in removed:
                continue

            seg = copy.copy(segments[i])
            seg.points = list(seg.points)

            # 端点辞書で隣接セグメントを探索し結合（結合後は端点が変わるので再探索）
            merged = True
            while merged:
                merged = False

                # 候補収集: seg の終点 == 候補の始点、または seg の始点 == 候補の終点
                candidates = set()
                for j in start_map.get((seg.fclass, seg.points[-1]), ()):
                    if j > i and j not in removed:
                        candidates.add(j)
                for j in end_map.get((seg.fclass, seg.points[0]), ()):
                    if j > i and j not in removed:
                        candidates.add(j)

                for j in sorted(candidates):
                    if j in removed:
                        continue
                    if not are_adjacent(seg, segments[j]):
                        continue

                    interpolate_attributes(seg, segments[j])

                    if seg.name == segments[j].name and seg.ref == segments[j].ref:
                        # 結合
                        if seg.points[-1] == segments[j].points[0]:
                            seg.points = seg.points + segments[j].points
                        else:
                            seg.points = segments[j].points + seg.points
                        print(f"combined {i} {j} {seg.fclass} {seg.ref} {seg.name}")
                        removed.add(j)
                        merged = True
                        break  # 端点が変わったので再探索

            new_segments.append(seg)

        print(f"{'#' * 80}\n{len(new_segments)}/{len(segments)}\n{'#' * 80}")
        if len(new_segments) == len(segments):
            break

        segments = new_segments

    return segments


def write_geojson(segments, output_dir):
    os.makedirs(output_dir, exist_ok=True)

    grouped = defaultdict(list)
    for seg in segments:
        grouped[seg.fclass].append(seg)

    written = []
    for fclass, segs in sorted(grouped.items()):
        features = []
        for seg in segs:
            features.append({
                "type": "Feature",
                "properties": {
                    "fclass": seg.fclass,
                    "name": seg.name,
                    "ref": seg.ref,
                },
                "geometry": {
                    "type": "LineString",
                    "coordinates": seg.points,
                },
            })

        geojson = {
            "type": "FeatureCollection",
            "features": features,
        }

        filename = f"osm_{fclass}.geojson"
        filepath = os.path.join(output_dir, filename)
        with open(filepath, "w", encoding="utf-8") as f:
            json.dump(geojson, f, ensure_ascii=False)

        gz_filepath = filepath + ".gz"
        with gzip.open(gz_filepath, "wt", encoding="utf-8") as f:
            json.dump(geojson, f, ensure_ascii=False)

        print(f"wrote {filepath} + .gz ({len(features)} features)")
        written.append(filepath)

    return written


def main():
	"""
	コマンドライン例：
	python .\osm_extract_simplify.py japan-260209.osm.pbf -o output -i 10
	"""
	
    parser = argparse.ArgumentParser(
        description="OSM PBF から主要道路を抽出・結合し GeoJSON で出力する")
    parser.add_argument("input", help="入力 .osm.pbf ファイル")
    parser.add_argument("-o", "--output", default=".", help="出力ディレクトリ (default: .)")
    parser.add_argument("-c", "--classes", nargs="+", default=TARGET_CLASSES,
                        help="対象道路種別 (default: motorway trunk primary secondary)")
    parser.add_argument("-i", "--max-iter", type=int, default=3,
                        help="結合イテレーション回数 (default: 3)")
    args = parser.parse_args()

    print("READING AND EXTRACTING...")
    segments = read_pbf(args.input, args.classes)
    print(f"extracted size: {len(segments)}")

    print("COMBINING...")
    segments = combine_segments(segments, max_iter=args.max_iter)

    print("WRITING...")
    write_geojson(segments, args.output)

    print("COMPLETED")


if __name__ == "__main__":
    main()

File: osm_extract_simplify.jl
#!/usr/bin/env julia
# OSM PBF から主要道路を抽出・結合し GeoJSON で出力する

using OpenStreetMapPBF
using JSON
using CodecZlib

# ─── データ構造 ───────────────────────────────────────────────

mutable struct RoadSegment
    fclass::String        # "motorway", "trunk", "primary", "secondary"
    name::String          # 道路名（""=未設定）
    ref::String           # 路線番号（""=未設定）
    points::Vector{Tuple{Float64,Float64}}  # [(lon, lat), ...]
end

const DEFAULT_CLASSES = ["motorway", "trunk", "primary", "secondary"]

# ─── PBF 読み込み（2パス） ────────────────────────────────────

struct RawWay
    fclass::String
    name::String
    ref::String
    node_ids::Vector{Int64}
end

function read_pbf(path::String, target_classes::Vector{String})
    targets = Set(target_classes)
    link_targets = Set(c * "_link" for c in target_classes)
    link_base = Dict(c * "_link" => c for c in target_classes)

    # Pass 1: Way を収集し、必要な node ID を集める
    raw_ways = RawWay[]
    needed_nodes = Set{Int64}()

    scan_ways(path) do way
        hw = get(way.tags, "highway", "")
        if hw in targets
            fclass = hw
        elseif hw in link_targets
            fclass = link_base[hw]
        else
            return
        end
        length(way.nodes) < 2 && return
        push!(raw_ways, RawWay(fclass,
                               get(way.tags, "name", ""),
                               get(way.tags, "ref", ""),
                               copy(way.nodes)))
        union!(needed_nodes, way.nodes)
    end

    # Pass 2: Node 座標を収集
    node_coords = Dict{Int64, Tuple{Float64,Float64}}()

    scan_nodes(path) do node
        if node.id in needed_nodes
            node_coords[node.id] = (node.lon, node.lat)
        end
    end

    # RoadSegment に変換
    segments = RoadSegment[]
    for rw in raw_ways
        pts = Tuple{Float64,Float64}[]
        valid = true
        for nid in rw.node_ids
            if haskey(node_coords, nid)
                push!(pts, node_coords[nid])
            else
                valid = false
                break
            end
        end
        if valid && length(pts) >= 2
            push!(segments, RoadSegment(rw.fclass, rw.name, rw.ref, pts))
        end
    end

    return segments
end

# ─── 属性補完 ─────────────────────────────────────────────────

function interpolate_attributes!(seg_i::RoadSegment, seg_j::RoadSegment)
    for (src, dst) in ((seg_i, seg_j), (seg_j, seg_i))
        if src.name != "" && src.name == dst.name
            if dst.ref == "" && src.ref != ""
                dst.ref = src.ref
            end
        end
        if src.ref != "" && src.ref == dst.ref
            if dst.name == "" && src.name != ""
                dst.name = src.name
            end
        end
    end
end

# ─── 隣接判定 ─────────────────────────────────────────────────

function are_adjacent(seg_a::RoadSegment, seg_b::RoadSegment)::Bool
    return seg_a.points[end] == seg_b.points[1] ||
           seg_a.points[1] == seg_b.points[end]
end

# ─── セグメント結合 ───────────────────────────────────────────

function combine_segments(segments::Vector{RoadSegment}, max_iter::Int=3)
    for iteration in 0:max_iter-1
        println("iteration $iteration")

        n = length(segments)

        # 端点からセグメントインデックスへの辞書を構築（O(n)）
        start_map = Dict{Tuple{String,Tuple{Float64,Float64}}, Set{Int}}()
        end_map   = Dict{Tuple{String,Tuple{Float64,Float64}}, Set{Int}}()
        for idx in 1:n
            seg = segments[idx]
            sk = (seg.fclass, seg.points[1])
            ek = (seg.fclass, seg.points[end])
            push!(get!(start_map, sk, Set{Int}()), idx)
            push!(get!(end_map,   ek, Set{Int}()), idx)
        end

        removed = Set{Int}()
        new_segments = RoadSegment[]

        for i in 1:n
            i in removed && continue

            seg = RoadSegment(segments[i].fclass, segments[i].name,
                              segments[i].ref, copy(segments[i].points))

            # 端点辞書で隣接セグメントを探索し結合（結合後は端点が変わるので再探索）
            merged = true
            while merged
                merged = false

                # 候補収集: seg の終点 == 候補の始点、または seg の始点 == 候補の終点
                candidates = Set{Int}()
                for j in get(start_map, (seg.fclass, seg.points[end]), ())
                    if j > i && !(j in removed)
                        push!(candidates, j)
                    end
                end
                for j in get(end_map, (seg.fclass, seg.points[1]), ())
                    if j > i && !(j in removed)
                        push!(candidates, j)
                    end
                end

                for j in sort!(collect(candidates))
                    j in removed && continue
                    are_adjacent(seg, segments[j]) || continue

                    interpolate_attributes!(seg, segments[j])

                    if seg.name == segments[j].name && seg.ref == segments[j].ref
                        # 結合（接続点の重複を除去）
                        if seg.points[end] == segments[j].points[1]
                            append!(seg.points, @view segments[j].points[2:end])
                        else
                            seg.points = vcat(segments[j].points, @view seg.points[2:end])
                        end
                        println("combined $i $j $(seg.fclass) $(seg.ref) $(seg.name)")
                        push!(removed, j)
                        merged = true
                        break  # 端点が変わったので再探索
                    end
                end
            end

            push!(new_segments, seg)
        end

        println("#"^80)
        println("$(length(new_segments))/$(length(segments))")
        println("#"^80)

        if length(new_segments) == length(segments)
            break
        end

        segments = new_segments
    end

    return segments
end

# ─── GeoJSON 出力 ─────────────────────────────────────────────

function write_geojson(segments::Vector{RoadSegment}, output_dir::String)
    mkpath(output_dir)

    grouped = Dict{String, Vector{RoadSegment}}()
    for seg in segments
        push!(get!(grouped, seg.fclass, RoadSegment[]), seg)
    end

    for fclass in sort(collect(keys(grouped)))
        segs = grouped[fclass]
        features = Vector{Dict{String,Any}}(undef, length(segs))

        for (idx, seg) in enumerate(segs)
            features[idx] = Dict{String,Any}(
                "type" => "Feature",
                "properties" => Dict{String,Any}(
                    "fclass" => seg.fclass,
                    "name" => seg.name,
                    "ref" => seg.ref,
                ),
                "geometry" => Dict{String,Any}(
                    "type" => "LineString",
                    "coordinates" => [[round(pt[1], digits=5), round(pt[2], digits=5)] for pt in seg.points],
                ),
            )
        end

        geojson = Dict{String,Any}(
            "type" => "FeatureCollection",
            "features" => features,
        )

        filepath = joinpath(output_dir, "osm_$(fclass).geojson")
        open(filepath, "w") do f
            JSON.print(f, geojson)
        end

        gz_filepath = filepath * ".gz"
        open(gz_filepath, "w") do f
            gzf = GzipCompressorStream(f)
            JSON.print(gzf, geojson)
            close(gzf)
        end

        println("wrote $filepath + .gz ($(length(features)) features)")
    end
end

# ─── メイン ───────────────────────────────────────────────────

# コマンドライン例：
# julia osm_extract_simplify.jl japan-260209.osm.pbf -o output -i 10

function main()
    if length(ARGS) < 1
        println(stderr, "Usage: julia osm_extract_simplify.jl INPUT.osm.pbf [-o DIR] [-c CLASS...] [-i N]")
        exit(1)
    end

    # 引数パース
    input_file = ""
    output_dir = "."
    classes = copy(DEFAULT_CLASSES)
    max_iter = 3

    i = 1
    while i <= length(ARGS)
        arg = ARGS[i]
        if arg == "-o" || arg == "--output"
            i += 1
            output_dir = ARGS[i]
        elseif arg == "-c" || arg == "--classes"
            classes = String[]
            i += 1
            while i <= length(ARGS) && !startswith(ARGS[i], "-")
                push!(classes, ARGS[i])
                i += 1
            end
            continue  # i は既に次の引数を指している
        elseif arg == "-i" || arg == "--max-iter"
            i += 1
            max_iter = parse(Int, ARGS[i])
        else
            input_file = arg
        end
        i += 1
    end

    if input_file == ""
        println(stderr, "Error: input file is required")
        exit(1)
    end

    println("READING AND EXTRACTING...")
    segments = read_pbf(input_file, classes)
    println("extracted size: $(length(segments))")

    println("COMBINING...")
    segments = combine_segments(segments, max_iter)

    println("WRITING...")
    write_geojson(segments, output_dir)

    println("COMPLETED")
end

main()
